TE-496 connect and disconnect

kbayraktar opened the pull request.

kbayraktar committed:
TE-495 rudimentary host-fixture initial commit

Commit:
Connect and disconnect implemented

Commit:
TE-496 spaces removed.

franzbecker commented on .gitignore:
```diff
@@ -0,0 +1,15 @@
+# Java build artifacts class files
```
Please provide a `.gitignore` that is the same as in the other fixtures. See https://github.com/test-editor/rest-fixture/blob/develop/.gitignore

franzbecker commented on .classpath:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
We don't want to check in Eclipse specific stuff in the fixtures. See comment to `.gitignore`

franzbecker commented on build.gradle:
```diff
@@ -0,0 +1,40 @@
+plugins { 
```
Please use a `build.gradle` that is similar to the existing ones. If each fixture starts writing its build.gradle from scratch they all differ. See https://github.com/test-editor/rest-fixture/blob/develop/build.gradle

franzbecker commented on settings.gradle:
```diff
@@ -0,0 +1,18 @@
+/*
```
Please remove those comments

franzbecker commented on src/main/java/org/testeditor/fixture/host/HostDriverFixture.aml:
```diff
@@ -0,0 +1,16 @@
+package org.testeditor.fixture.host
+
+interaction type connect {
+    template = "Connect to " ${hostname} " with path " s3270Path " and port " ${port}  
```
The spaces in the template are not necessary. Please write `"Connect to" ${hostname} "with path"`

gunther-bachmann replied:
should be ${s3270Path}  right?

franzbecker commented on src/main/java/org/testeditor/fixture/host/HostDriverFixture.java:
```diff
@@ -0,0 +1,53 @@
+import org.testeditor.fixture.host.s3270.options.TerminalMode;
+import org.testeditor.fixture.host.s3270.options.TerminalType;
+
+public class HostDriverFixture {
```
Please use 4 spaces instead of 2 spaces as formatting.

kbayraktar replied:
We should use Workspace Mechanics for harmonizing our code for the whole team members.

franzbecker commented on src/main/java/org/testeditor/fixture/host/HostDriverFixture.java:
```diff
@@ -0,0 +1,53 @@
+public class HostDriverFixture {
+  Connection con;
+
+  Logger logger = LoggerFactory.getLogger(HostDriverFixture.class);
```
should be `private static final` and above the `Connection` variable

franzbecker commented on src/main/java/org/testeditor/fixture/host/HostDriverFixture.java:
```diff
@@ -0,0 +1,53 @@
+import org.testeditor.fixture.host.s3270.options.TerminalType;
+
+public class HostDriverFixture {
+  Connection con;
```
Should be `private` and named `connection`

franzbecker commented on src/main/java/org/testeditor/fixture/host/HostDriverFixture.java:
```diff
@@ -0,0 +1,53 @@
+   * @param hostname
+   * @param port
+   */
+  public void connect(String s3270Path, String hostname, int port) {
```
Method needs to be annotated with `@FixtureMethod` in order to be seen by the test-editor.

franzbecker commented on src/main/java/org/testeditor/fixture/host/HostDriverFixture.java:
```diff
@@ -0,0 +1,53 @@
+   * @param port
+   */
+  public void connect(String s3270Path, String hostname, int port) {
+    logger.info("Host-Fixture starting ...");
```
A better log message would be `logger.info("Connecting to host='{}', port='{}' with s3270path='{}'", hostname, port, s3270path);`

kbayraktar replied:
The connection method does that in the Connection Class when the s3270 process starts.

franzbecker commented on src/main/java/org/testeditor/fixture/host/HostDriverFixture.java:
```diff
@@ -0,0 +1,53 @@
+    if (connected) {
+      logger.info("Host-Fixture ready to rock ;O)");
+    } else {
+      logger.info("Host-Fixture is not ready to rock :O(");
```
We should fail here, i.e. throw an exception

franzbecker commented on src/main/java/org/testeditor/fixture/host/HostDriverFixture.java:
```diff
@@ -0,0 +1,53 @@
+    con.connect(s3270Path, hostname, port, type, mode, charSet);
+    boolean connected = con.isConnected();
+    if (connected) {
+      logger.info("Host-Fixture ready to rock ;O)");
```
Better `successfully connected to host='{}', port='{}'`

franzbecker commented on src/main/java/org/testeditor/fixture/host/HostDriverFixture.java:
```diff
@@ -0,0 +1,53 @@
+  }
+
+  public void disconnect() {
+    logger.info("Connection closing ...");
```
log message should appear after the logout and say something like "Successfully disconnected".

kbayraktar replied:
This is also done by the Connection

franzbecker commented on src/main/java/org/testeditor/fixture/host/HostDriverFixture.java:
```diff
@@ -0,0 +1,53 @@
+    con.disconnect();
+  }
+
+  public void getStatus() {
```
What is this used for? There is no return value.

kbayraktar replied:
The status will be handled in TE-497, the method getStatus() and typeInto() are connected to the not yet implemented methods with  "// TODO is coming next ;O)" statements in connection.getStatus() and connection.typeInto(). -> coming soon.

franzbecker commented on src/main/java/org/testeditor/fixture/host/HostDriverFixture.java:
```diff
@@ -0,0 +1,53 @@
+  }
+
+  public void typeInto(String value, int row, int col) {
+    logger.info("Set cursor position ...");
```
Log message should appear after the position is set and include the values.

kbayraktar replied:
same as above :-) Status will be implemented in next PullRequest same as typeInto() -> coming soon.

franzbecker commented on src/main/java/org/testeditor/fixture/host/Locator.java:
```diff
@@ -0,0 +1,49 @@
+
+  /**
+   *
+   * @param rol
```
useless javadoc in this class, please delete

franzbecker commented on src/main/java/org/testeditor/fixture/host/Locator.java:
```diff
@@ -0,0 +1,49 @@
+   * @param row
+   */
+  public Locator(int col, int row) {
+
```
empty line

franzbecker commented on src/main/java/org/testeditor/fixture/host/io/ErrorReader.java:
```diff
@@ -0,0 +1,50 @@
+  @Override
+  public void run() {
+
+    err = new BufferedReader(new InputStreamReader(process.getErrorStream()));
```
I don't see why the assignment to the instance variable is necessary here.

franzbecker commented on src/main/java/org/testeditor/fixture/host/io/ErrorReader.java:
```diff
@@ -0,0 +1,50 @@
+
+    err = new BufferedReader(new InputStreamReader(process.getErrorStream()));
+    try {
+      while (true) {
```
There are better approaches of reading an input stream into a String.

franzbecker commented on src/main/java/org/testeditor/fixture/host/io/ErrorReader.java:
```diff
@@ -0,0 +1,50 @@
+import java.io.IOException;
+import java.io.InputStreamReader;
+
+public class ErrorReader extends Thread {
```
This class is never used so far.

franzbecker commented on src/main/java/org/testeditor/fixture/host/net/Connection.java:
```diff
@@ -0,0 +1,222 @@
+    String commandLine = String.format("%s -charset %s -model %s-%d %s:%d -utf8", s3270Path,
+        charSet.getCharSet(), type.getType(), mode.getMode(), hostname, port);
+    try {
+      logger.info("starting " + commandLine);
```
This is unreadable by a human without knowing the source code. There should be a better output for the user! Also the log level should be debug.

kbayraktar replied:
Totally agree with u. This will be implemented in TE-497

franzbecker commented on src/main/java/org/testeditor/fixture/host/net/Connection.java:
```diff
@@ -0,0 +1,222 @@
+      return this;
+
+    } catch (final IOException ex) {
+      throw new RuntimeException("IO Exception while starting s3270/ws370.exe", ex);
```
The `s3270Path` would be useful to pass here instead of `s3270/ws370.exe`.

kbayraktar replied:
?

franzbecker commented on src/main/java/org/testeditor/fixture/host/net/Connection.java:
```diff
@@ -0,0 +1,222 @@
+    doQuit();
+    stop2370Process();
+    try {
+      s3270Process.waitFor();
```
We could also use `waitFor(long timeout, TimeUnit unit)` here

franzbecker commented on src/main/java/org/testeditor/fixture/host/net/Connection.java:
```diff
@@ -0,0 +1,222 @@
+  /**
+   * Disconnect from mainframe. Destroys all opened Input- and OutputStreams.
+   */
+  public void disconnect() {
```
The disconnect looks a bit confusing to me and should be rewritten. First, a quit command is sent to the process, afterwards it is destroyed and finally waited for its termination. A better way would be to send the quit command, wait for a certain timeout and if that doesn't occur kill the process. Also the cleanup logic is partly in this method, partly in the `cleanup()` method.

franzbecker commented on src/main/java/org/testeditor/fixture/host/net/Connection.java:
```diff
@@ -0,0 +1,222 @@
+   * stop the s3270 process
+   */
+  private void stop2370Process() {
+    new Thread(new Runnable() {
```
I don't see why we need a separate thread here. Also with the comment above the sleeping might not be necessary.

franzbecker commented on src/main/java/org/testeditor/fixture/host/net/Connection.java:
```diff
@@ -0,0 +1,222 @@
+          if (s3270Process != null) {
+            s3270Process.destroy();
+          }
+        } catch (final InterruptedException ex) {
```
The `InterruptedException` is thrown because of the sleeping. There is no need to duplicate the `destroy()` call in the catch block.

franzbecker commented on src/main/java/org/testeditor/fixture/host/net/Connection.java:
```diff
@@ -0,0 +1,222 @@
+      return false;
+    } else {
+      final Result r = doCommand("");
+      if (r.getStatus().matches(". . . C.*")) {
```
There should be a proper abstraction for the status String.

franzbecker commented on src/main/java/org/testeditor/fixture/host/net/Connection.java:
```diff
@@ -0,0 +1,222 @@
+      if (r.getStatus().matches(". . . C.*")) {
+        return true;
+      } else {
+        out.println("quit");
```
That's a bit suprising and a sideeffect. Why is the connection killed when the process is no longer connected?

franzbecker commented on src/main/java/org/testeditor/fixture/host/net/Connection.java:
```diff
@@ -0,0 +1,222 @@
+    }
+  }
+
+  private void readOutput(List<String> lines) throws IOException {
```
The list should not be passed in as a parameter but be created by this method.

franzbecker replied:
Also why not use `in.lines().collect(Collectors::toList)`? logging should be done separately if the log level is debug.

franzbecker commented on src/main/java/org/testeditor/fixture/host/net/Connection.java:
```diff
@@ -0,0 +1,222 @@
+      out.flush();
+      logger.debug(
+          "*****************************************************************************************");
+      logger.debug("---> \"{}\" entered ", command);
```
I would prefer: `("Command sent: {}", command`)

franzbecker commented on src/test/java/org/testeditor/fixture/host/HostDriverFixtureTest.java:
```diff
@@ -0,0 +1,19 @@
+public class HostDriverFixtureTest {
+
+  @Test
+  public void connectionTest() {
```
That is a manual integration test, not a unit test.

kbayraktar replied:
Yes we should discuss, how to test the host-fixture completely.

franzbecker commented on src/test/resources/log4j2.xml:
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Should use the same configuration as the other fixtures, see https://github.com/test-editor/rest-fixture/blob/develop/src/test/resources/log4j2.xml

franzbecker reviewed (changes requested):
The code needs some cleanup as indicated by the comments. Also a Jenkinsfile and unit tests are missing.

Commit:
TE-496 pull request changes

Commit:
TE-496 pull request changes

Commit:
TE-496 pull request changes

Commit:
TE-496 Modultests added
-  changed Mockito to version 2.77.22
-  Integrationtest added (but only for a special environment)
- the methods connect() and disconnect() returns boolean now

franzbecker commented on settings.gradle:
```diff
@@ -0,0 +1 @@
+rootProject.name = 'hostfixture'
```
should be `host-fixture`

franzbecker commented on src/main/java/org/testeditor/fixture/host/HostDriverFixture.aml:
```diff
@@ -0,0 +1,16 @@
+package org.testeditor.fixture.host
+
+interaction type connect {
+    template = "Connect to" ${hostname} "with path" s3270Path "and port" ${port}  
```
I think hostname and port should be closer together, shouldn't they?
How about `"Connect to" ${hostname} "with port" ${port} "using s3270 from" ${s3270Path}`

franzbecker commented on src/main/java/org/testeditor/fixture/host/HostDriverFixture.java:
```diff
@@ -0,0 +1,61 @@
+    @FixtureMethod
+    public boolean disconnect() {
+        logger.info("Connection closing ...");
+        return !connection.disconnect();
```
Should not be negated. `connection.disconnect()` should return true when the disconnect is successful.

franzbecker commented on src/main/java/org/testeditor/fixture/host/Locator.java:
```diff
@@ -0,0 +1,43 @@
+  }
+
+  /**
+   * @return the col
```
useless comment, same as the other ones below

franzbecker replied:
useless comment, same as the other ones below

franzbecker commented on src/main/java/org/testeditor/fixture/host/net/Connection.java:
```diff
@@ -0,0 +1,225 @@
+                    }
+                } catch (final InterruptedException ex) {
+                    if (s3270Process != null) {
+                        s3270Process.destroy();
```
duplicate here. The `InterruptedException` is only for the sleeping so it should wrap just the `Thread.sleep(...)`

franzbecker commented on src/test/java/org/testeditor/fixture/host/HostDriverFixtureTest.java:
```diff
@@ -0,0 +1,24 @@
+
+        HostDriverFixture hdf = mock(HostDriverFixture.class);
+        when(hdf.connect(S2370_PATH, hostUrl, hostPort)).thenReturn(true);
+        assertEquals(hdf.connect(S2370_PATH, hostUrl, hostPort), true);
```
This just tests the mock.

franzbecker commented on src/test/java/org/testeditor/fixture/host/integrationtest/HostDriverIntegrationTest.java:
```diff
@@ -0,0 +1,27 @@
+
+public class HostDriverIntegrationTest {
+
+    // @Test
```
`@Test` is ok but you should add `@Ignore` instead - e.g. `@Ignore("manual execution only")`

Commit:
TE-496 pull request changes added

gunther-bachmann commented on src/main/java/org/testeditor/fixture/host/HostDriverFixture.java:
```diff
@@ -0,0 +1,59 @@
+     */
+    @FixtureMethod
+    public boolean connect(String s3270Path, String hostname, int port) {
+        logger.info("Host-Fixture starting ...");
```
how about "connecting" instead of "starting"

gunther-bachmann commented on src/main/java/org/testeditor/fixture/host/HostDriverFixture.java:
```diff
@@ -0,0 +1,59 @@
+    @FixtureMethod
+    public boolean disconnect() {
+        logger.info("Connection closing ...");
+        return connection.disconnect();
```
failing disconnect should raise exception (as connect does in case of failure)

gunther-bachmann commented on src/main/java/org/testeditor/fixture/host/net/Connection.java:
```diff
@@ -0,0 +1,227 @@
+            logger.error("Something went wrong during closing InputStreamreader of s2370 process");
+        }
+        cleanup();
+        logger.info("Disconnected successfully from host : {} ", hostname);
```
successful even though I do not know whether isConnected is still true?

gunther-bachmann commented on src/main/java/org/testeditor/fixture/host/net/Connection.java:
```diff
@@ -0,0 +1,227 @@
+            @Override
+            public void run() {
+                try {
+                    Thread.sleep(1000);
```
only sleep?

gunther-bachmann commented on src/main/java/org/testeditor/fixture/host/net/Connection.java:
```diff
@@ -0,0 +1,227 @@
+            if (r.getStatus().matches(". . . C.*")) {
+                return true;
+            } else {
+                out.println("quit");
```
code duplication with doQuit?

gunther-bachmann commented on src/main/java/org/testeditor/fixture/host/net/Connection.java:
```diff
@@ -0,0 +1,227 @@
+                out.println("quit");
+                out.flush();
+                s3270Process.destroy();
+                s3270Process = null;
```
code duplication with cleanup?

gunther-bachmann commented on src/main/java/org/testeditor/fixture/host/net/Connection.java:
```diff
@@ -0,0 +1,227 @@
+            logger.debug("*****************************************************************************************");
+            logger.debug("---> Command sent: \"{}\"", command);
+            List<String> lines = new ArrayList<String>();
+            readOutput(lines);
```
why not return lines by calling readOutput ?

gunther-bachmann commented on src/main/java/org/testeditor/fixture/host/net/Connection.java:
```diff
@@ -0,0 +1,227 @@
+            readOutput(lines);
+            int size = lines.size();
+            if (size > 0) {
+                Result result = new Result(lines.subList(0, size - 1), lines.get(size - 1));
```
why drop the list line? maybe put it into a separate method that describes why this is done?

kbayraktar replied:
#ignore

gunther-bachmann commented on src/main/java/org/testeditor/fixture/host/net/Connection.java:
```diff
@@ -0,0 +1,227 @@
+        }
+    }
+
+    private void processAvailable() {
```
name suggests that bool is returned. Maybe use different name (e.g. assertProcessAvailable) ...

gunther-bachmann commented on src/main/java/org/testeditor/fixture/host/s3270/Result.java:
```diff
@@ -0,0 +1,25 @@
+ */
+public class Result {
+
+  private final List<String> data;
```
why not name this "lines"?

Commit:
TE-496 pull request changes incl. ModulTests for HostFixture
